fdb-extensions/src/test/java/com/apple/foundationdb/async/hnsw/HNSWTest.java

```diff
@@ -0,0 +1,542 @@
+
+        final ImmutableSet<Tuple> trueNN =
+                recordsOrderedByDistance.stream()
+                        .map(PrimaryKeyVectorAndDistance::getPrimaryKey)
+                        .collect(ImmutableSet.toImmutableSet());
+
+        int recallCount = 0;
+        for (ResultEntry resultEntry : results) {
+            logger.info("nodeId ={} at distance={}", resultEntry.getPrimaryKey().getLong(0),
+                    resultEntry.getDistance());
+            if (trueNN.contains(resultEntry.getPrimaryKey())) {
+                recallCount ++;
+            }
+        }
+        final double recall = (double)recallCount / (double)k;
+        logger.info("search transaction took elapsedTime={}ms; read nodes={}, read bytes={}, recall={}",
+                TimeUnit.NANOSECONDS.toMillis(endTs - beginTs),
+                onReadListener.getNodeCountByLayer(), onReadListener.getBytesReadByLayer(),
+                String.format(Locale.ROOT, "%.2f", recall * 100.0d));
+        Assertions.assertThat(recall).isGreaterThan(0.79);
+
+        final Set<Long> insertedIds =
+                LongStream.range(0, 1000)
+                        .boxed()
+                        .collect(Collectors.toSet());
+
+        final Set<Long> readIds = Sets.newHashSet();
+        hnsw.scanLayer(db, 0, 100,
+                node -> Assertions.assertThat(readIds.add(node.getPrimaryKey().getLong(0))).isTrue());
+        Assertions.assertThat(readIds).isEqualTo(insertedIds);
+
+        readIds.clear();
+        hnsw.scanLayer(db, 1, 100,
+                node -> Assertions.assertThat(readIds.add(node.getPrimaryKey().getLong(0))).isTrue());
+        Assertions.assertThat(readIds.size()).isBetween(10, 50);
+    }
+
+    private int basicInsertBatch(final HNSW hnsw, final int batchSize,
+                                 @Nonnull final AtomicLong nextNodeIdAtomic, @Nonnull final TestOnReadListener onReadListener,
+                                 @Nonnull final Function<Transaction, PrimaryKeyAndVector> insertFunction) {
+        return db.run(tr -> {
+            onReadListener.reset();
+            final long nextNodeId = nextNodeIdAtomic.get();
+            final long beginTs = System.nanoTime();
+            for (int i = 0; i < batchSize; i ++) {
+                final var record = insertFunction.apply(tr);
+                if (record == null) {
+                    return i;
+                }
+                hnsw.insert(tr, record.getPrimaryKey(), record.getVector()).join();
+            }
+            final long endTs = System.nanoTime();
+            logger.info("inserted batchSize={} records starting at nodeId={} took elapsedTime={}ms, readCounts={}, readBytes={}",
+                    batchSize, nextNodeId, TimeUnit.NANOSECONDS.toMillis(endTs - beginTs),
+                    onReadListener.getNodeCountByLayer(), onReadListener.getBytesReadByLayer());
+            return batchSize;
+        });
+    }
+
+    @Test
+    @SuperSlow
```

Comment:
Does this have to be done in `SuperSlow` mode? Is there a variant where we load, say, a subset of the data set and we're able to do that in normal tests?

Reply:
This test takes exactly 3min10sec to run. It should be fine to not run it under `@SuperSlow`. If you feel like that's too long, there is not much we can do as this is a real dataset and this is its smallest size. We could filter this dataset to something even smaller but then we would have to host it.

Reply:
Update, it is not fine. I will put it under `@SuperSlow` again. I think the existing coverage without this always being enabled should be enough, though.